pkg/planner/core/find_best_task.go

```diff
@@ -1567,6 +1567,32 @@ func (ds *DataSource) FindBestTask(prop *property.PhysicalProperty, planCounter
 	return
 }
 
+func (ds *DataSource) compareTaskRegardingStore(curTask, bestTask base.Task, curTaskStore, bestTaskStore kv.StoreType,
+	op *optimizetrace.PhysicalOptimizeOp) (curIsBetter bool, err error) {
+	if bestTask.Invalid() { // in this case the curTask is always better to let us at least can get a valid plan.
+		return true, nil
+	}
+
+	switch ds.preferStoreType { // consider the storage preference
+	case h.PreferTiFlash:
+		if bestTaskStore == kv.TiFlash && curTaskStore == kv.TiKV {
```

Comment:
```suggestion
		if bestTaskStore == kv.TiFlash && curTaskStore != kv.TiFlash {
```
`curTaskStore` could be other types, like `kv.TiDB`.
The same below.